### DISCUSSION TITLE
primitive data인데 어떻게 내장 함수를 사용할 수 있는걸까요?

### DISCUSSION THREAD

**Discussion by kamwoo** [woowacourse-study/2022-woowahan-bansanghwe#109] [JavaScript]
자바스크립는 크게 3가지의 객체로 분류할 수 있다.

- 표준 빌트인 객체
- 호스트 객체
- 사용자 정의 객체

### 표준 빌트인 객체

ECMAScript spec에 정의된 객체를 말하며 전역에 공통 기능으로 제공된다. 자바스크립트 실행 환경에 관계없이 사용할 수 있고, 전역 객체의 프로퍼티로서 제공된다. 

Object, String, Number, Date, Array, Function 등으로 약 40개 정도의 표준 빌트인 객체가 있다.

대부분의 빌트인 객체는 인스턴스를 생성할 수 있는 생성자 함수 객체로 프로토타입 메서드와 정적 메서드를 제공한다.

```tsx
const strObj = new String('kam'); // String {"kam"}
const numObj = new Number(123) // Number {123}
```

빌트인 객체가 생성한 인스턴스의 프로토타입은 빌트인 객체의 프로토타입에 바인딩된다.

### 호스트 객체

ECMAScript spec에 정의되어 있지 않지만 자바스크립트 환경에서 제공하는 객체를 말한다. 브라우저 환경에서 DOM, fetch, RequestAnimationFrame, WebStorage 등이 해당된다.

### 사용자 정의 객체

사용자가 직접 정의한 객체를 말한다.

## **primitive data인데 어떻게 내장 함수를 사용할 수 있는걸까?**

원시값인 문자열, 숫자 등과 같은 경우 객체처럼 `.` 을 통해서 접근하게 되면 자바스트립트 엔진이 일시적으로 원시값을 연관된 객체로 변환해주기 때문이다. 

원시값을 객체처럼 사용하면 자바스크립트 엔진이 암묵적으로 연관된 객체를 생성하여 생성된 객체로 프로퍼티에 접근하거나 메서드를 호출하고 다시 원시값으로 되돌린다.

이처럼 잠시 생성되는 임시 객체를 Wrapper Object라고 한다.

예를 들어서 string 원시값을 객체처럼 접근하게 되면

```tsx
const str = 'hello'; // 1
const str.name = 'kam'; // 2
	               //3
console.log(str.name); // 4
console.log(typeof str, str); // 5
```

1. str 변수는 ‘hello’ 문자열을 값으로 가진다.
2. 원시값인 문자열에 객체처럼 접근해서 ‘name’프로터티에 ‘kam’을 저장하려고한다.
    1. 이때 str 변수는 임시로 생성된 String 래퍼 객체를 가리킨다.
    2. ‘hello’는 래퍼 객체의 [[StringData]] 내부 슬롯에 할당된다.
    3. 래퍼 객체의 ‘name’ 프로퍼티가 동적으로 추가된다.
    4. ‘name’프로퍼티에 ‘kam’이 저장된다.
3. 변수 str은 래퍼 객체의 [[StringData]] 내부 슬롯에 할당된 원시값을 다시 가진다.
    
    래퍼 객체는 참조되지 않으므로 GC에 대상이된다.
    
4. str에 ‘name’프로퍼티에 접근하려고 한다.
    1. 이때 다시 새로운 String 래퍼객체를 임시로 생성한다.
    2. ‘hello’는 래퍼 객체의 [[StringData]] 내부 슬롯에 할당된다.
    3. 새로운 래퍼객체에는 ‘name’ 프로퍼티가 없으므로 ‘undefined’가 로그가 적힌다.
5. 3번 반복
6. 로그에 `String hello` 가 적힌다. 

모든 원시값이 해당하는 것은 아니고 number, string, boolean이 래퍼객체를 생성하고 null, undefined는 래퍼객체를 생성하지 않는다.

이렇게 사용하는 이유는 편리함 때문이라고 생각한다. 이처럼 래퍼객체로 인해서 원시값을 마치 객체처럼 사용해서 빌트인 객체의 프로토타입 메소드 또는 프로퍼티를 참조할 수 있기 때문에 new 연사자와 함께 생성자 함수를 호출해서 인스턴스를 생성할 필요가 없어지기 때문이다.

**Comment by airman5573** [comment]
어케했노..

**Comment by byhhh2** [comment]
원시값과 참조값이란?

**Comment by byhhh2** [comment]
undefined vs null

**Comment by byhhh2** [comment]
자바스크립트에서 객체란?

**Comment by byhhh2** [comment]
자바스크립트 언어는 왜 객체 지향이라고 불릴까?

**Comment by byhhh2** [comment]
자바스크립트에서 프로토타입이란?

**Comment by byhhh2** [comment]
다른 언어와 js(프로토타입)의 상속에 있어서 차이는?

**Comment by byhhh2** [comment]
자바스크립트에서 클래스란?

**Comment by byhhh2** [comment]
왜 클래스라는 문법을 도입했을까?

**Comment by byhhh2** [comment]
타입스크립트 왜 이용했나?

**Comment by byhhh2** [comment]
interface vs type

**Comment by byhhh2** [comment]
타입스크립트는 구조적 서브 타이핑 시스템을 갖는데 왜 추가적인 속성이 들어오면 오류를 띄울까?

**Comment by byhhh2** [comment]
any는 왜 지향해야 할까?

**Comment by byhhh2** [comment]
extends vs implements

**Comment by byhhh2** [comment]
자바스크립트의 메서드란?

**Comment by byhhh2** [comment]
메서드 vs 일반함수

**Comment by byhhh2** [comment]
메서드를 생성자 함수로 쓸 수 있나?

**Comment by byhhh2** [comment]
자바스크립트에서 스코프란?

**Comment by byhhh2** [comment]
스코프 체인이란?

**Comment by byhhh2** [comment]
스코프 체인과 프로토타입 체인의 관계는?

**Reply by byhhh2** [reply]
프로젝트에서 사용한 예시

**Comment by byhhh2** [comment]
typeof vs instanceof
